conjure-go-client/httpclient/body.go

```diff
@@ -0,0 +1,133 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpclient
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// RequestBody is an interface that can be used to set the body of an http.Request.
+// Implementations of this interface should set the fields Body, GetBody, and ContentLength (if known).
+type RequestBody interface {
+	setRequestBody(req *http.Request) error
+}
+
+// requestBodyFunc is a function that sets the body of an http.Request and implements RequestBody.
+type requestBodyFunc func() (length int64, body io.ReadCloser, getBody func() (io.ReadCloser, error), err error)
+
+func (f requestBodyFunc) setRequestBody(req *http.Request) (err error) {
+	req.ContentLength, req.Body, req.GetBody, err = f()
+	return err
+}
+
+// RequestBodyInMemory sets the *http.Request Body field to the provided *bytes.Buffer, *bytes.Reader, or *strings.Reader for upload.
+// The GetBody field is set to a function that returns the same io.ReadCloser.
+func RequestBodyInMemory[T bytes.Buffer | bytes.Reader | strings.Reader](input *T) RequestBody {
+	return requestBodyFunc(func() (int64, io.ReadCloser, func() (io.ReadCloser, error), error) {
+		if input == nil {
+			return 0, nil, nil, nil
+		}
+		contentLen := int64(any(input).(interface{ Len() int }).Len())
+		snapshot := *input
+		getBody := func() (io.ReadCloser, error) {
+			r := snapshot
+			return io.NopCloser(any(&r).(io.Reader)), nil
+		}
+		firstBody, _ := getBody()
+		return contentLen, firstBody, getBody, nil
+	})
+}
+
+type requestBodyStreamInput interface {
+	func() io.ReadCloser | func() (io.ReadCloser, error) | func() (io.ReadCloser, int64, error)
+}
+
+// RequestBodyStreamOnce sets the *http.Request Body field for upload.
+//
+// The GetBody field is left nil. The http.Transport will return an error if it is unable to replay the request body.
+// Use this function when the body should be read only once (e.g. it is a response body stream from another request).
+//
+// The body's Close() method will be called when the request is completed. To disable, wrap the value in io.NopCloser.
+//
+// The input argument must be a function of one of the following types:
+//   - func() io.ReadCloser                 // Returns the body and no error
+//   - func() (io.ReadCloser, error)        // Returns the body and an error
+//   - func() (io.ReadCloser, int64, error) // Returns the body, content length, and an error
+func RequestBodyStreamOnce[T requestBodyStreamInput](input T) RequestBody {
```

Comment:
Potential problem: This does not protect against CGR's own retry logic calling the provided input function multiple times to populate req.Body.

Should we wrap this in something stateful enough that it refuses and returns an error if the reader has already been read?

Reply:
Yes I think we should. The most confusing failure mode that exists today is CGR fails the first request for some reason (the request times out, for example), then proceeds with a retry. The error that users get back is the error from one of the retries - usually something opaque like `file already closed` or something like that.

Reply:
ack, I'll figure something out

Reply:
Ok I don't think it's my finest work, but I now have RequestBodyStreamOnce correctly refusing to retry